File: LinkedList/148_SortList.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
# @Author: xuezaigds@gmail.com
# @Last Modified time: 2016-09-06 20:03:53


# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# Merge Sort
class Solution(object):
    def sortList(self, head):
        if not head or not head.next:
            return head
        # Get the two half parts.
        pre_slow = None
        slow = fast = head
        while fast and fast.next:
            pre_slow = slow
            slow = slow.next
            fast = fast.next.next
        pre_slow.next = None        # Cut the linked list to two parts.
        left_list = self.sortList(head)
        right_list = self.sortList(slow)
        return self.merge(left_list, right_list)

    # Operator merge.
    def merge(self, left_list, right_list):
        pre_head = dummy = ListNode(None)
        while left_list and right_list:
            if left_list.val < right_list.val:
                dummy.next = left_list
                left_list = left_list.next
            else:
                dummy.next = right_list
                right_list = right_list.next
            dummy = dummy.next

        dummy.next = left_list or right_list

        return pre_head.next


# Quick sort:  Time Limit Exceeded
class Solution_2(object):
    def partition(self, begin, end):
        if not begin or begin.next == end:
            return begin
        pivot = begin.val
        keep, pos = begin, begin
        scan = begin.next
        while scan != end:
            if scan.val <= pivot:
                pos = pos.next
                if scan != pos:
                    scan.val, pos.val = pos.val, scan.val
            scan = scan.next

        pos.val, keep.val = keep.val, pos.val
        return pos

    def quick_sort(self, begin, end):
        if begin == end or begin.next == end:
            return begin

        pos = self.partition(begin, end)
        head = self.quick_sort(begin, pos)
        self.quick_sort(pos.next, end)
        return head

    def sortList(self, head):
        return self.quick_sort(head, None)

"""
[]
[1]
[1,2]
[5,1,2]
[5,1,2,3]
[5,1,2,3,6,7,8,9,12,2]
"""

File: LinkedList/148_SortList.cpp
/*
 * @Author: xuezaigds@gmail.com
 * @Last Modified time: 2016-09-06 20:40:19
 */

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

class Solution {
public:
    ListNode * merge_list(ListNode *left, ListNode *right){
        ListNode *pre_head = new ListNode(0);
        ListNode *scan = pre_head;
        while (left && right){
            if (left->val <= right->val){
                scan->next = left;
                left = left->next;
            }
            else{
                scan->next = right;
                right = right->next;
            }
            scan = scan->next;
        }
        if(left != nullptr) scan->next = left;
        if(right != nullptr) scan->next = right;
        return pre_head->next;
    }

    ListNode* sortList(ListNode* head) {
        if(head == nullptr || head->next == nullptr){
            return head;
        }
        ListNode *slow=head, *fast=head, *pre_tail= nullptr;
        // Cut the list into two parts
        while (fast != nullptr and fast->next != nullptr){
            pre_tail = slow;
            slow = slow->next;
            fast = fast->next->next;
        }
        pre_tail->next = nullptr;

        ListNode *left = sortList(head);
        ListNode *right = sortList(slow);
        return merge_list(left, right);
    }
};

// QuickSort.  Time Limit Exceeded if using no trick.
// Refer to:
// https://discuss.leetcode.com/topic/15029/56ms-c-solutions-using-quicksort-with-explanations/2
class Solution_2 {
public:
    ListNode* partition(ListNode *begin, ListNode *end){
        if(begin == nullptr || begin->next == end){
            return begin;
        }
        ListNode *scan = begin->next;
        int pivot = begin->val;
        ListNode *pos = begin;
        while(scan != end){
            if(scan->val <= pivot){
                pos = pos->next;
                if(pos != scan){
                    swap(pos->val, scan->val);
                }
            }
            scan = scan->next;
        }
        swap(pos->val, begin->val);
        return pos;
    }

    ListNode* quicksort(ListNode* begin, ListNode* end){
        if(begin == end || begin->next == end){
            return begin;
        }
        ListNode *pos = partition(begin, end);
        ListNode* head = quicksort(begin, pos);
        quicksort(pos->next, end);
        return head;
    }

    ListNode* sortList(ListNode* head) {
        return quicksort(head, nullptr);
    }
};

/*
[]
[1]
[1,2]
[5,1,2]
[5,1,2,3]
[5,1,2,3,6,7,8,9,12,2]
*/
